### PR TITLE
Sequencing groups update

### PR DESCRIPTION
**Pull request by vivbak**
Changes described in full here https://docs.google.com/document/d/1FxpsF5LfZZ5ZXWZL-JtELxPoToXMzRBrPjxyAYOba_k/edit?usp=sharing

- [x] Sample -> SequenceGroup
- [x] Sequence -> Assay 
- [x] Variable Clean up 

RestAPI Calls -> GraphQL 

- [x] Populate SequenceGroups
- [x] Populate Assays 
- [x] Populate Analysis 
- [x] Populate Participants 
- [x] General refactor inline with above. 

Testing 
- [x] cohort population 

A couple of things have deliberately been excluded from this PR:
- Filenames that are created have not been renamed, for example `sample_qc.ht`. 
- Parameter names that feed into external tools. For example, `--sample-name-map` is a parameter for `gatk GenomicsDBImport`. In this case, the `sample_map` variable that's passed here has not been renamed. 

There are several reasons for the above omissions, that can be discussed further in a separate forum e.g. Are these filename changes going to impact any downstream analyses? When we use 'sg' centred language, and other tools use 'sample' centred language, when does it make sense to draw the line? etc.
But most importantly, this PR is already  chonky enough, and we can functionally draw the line here. 

Also, the tests are not as clean as they could be, to be refactored in a subsequent PR. 

**Test Run**
https://batch.hail.populationgenomics.org.au/batches/424753/jobs/1

**Review Roster**
Hello valued reviewers, thank you for taking the time to look at this PR. Please don't be alarmed by the size, as the vast majority of the changes are trivial. 

Please be sure to create issues for changes that fall outside of the scope of the PR rather than requesting changes here. We're under a bit of a crunch to get this out. 

You are of course welcome to review the PR in it's entirety, but I've allocated everyone specific files to look at. 
| Reviewer  | Content |
| ------------- | ------------- |
| @daniaki   | /test, mt_to_es.py, benchmark_combiner.py, benchmark_metamist_read_pedigree.py   |
| @jmarshall   | batch.py, filetypes.py, inputs.py, resources.py, utils.py, workflow.py, status.py, /large_cohort, /stages (excluding gatk_sv)  | 
| @MattWellie  | /jobs, main.py, /stages/gatk* |
| @illusional | metamist.py, targets.py, populate_metamist.py, setup.py |
| @silkm  | README.MD, /configs (i.e. all the tomls)|

Thank you all in advance for your help!

### CODE REVIEW

File: cpg_workflows/metamist.py [hunk 1]
```diff
@@ -83,7 +83,7 @@
 
 GET_PEDIGREE_QUERY = gql(
     """
-        query MyQuery($metamist_proj: String!){
+        query PedgireeQuery($metamist_proj: String!){
```
Comment (MattWellie) [inline cpg_workflows/metamist.py:86]: PedigreeQuery!

Comment (vivbak) [reply cpg_workflows/metamist.py:86]: much prefer the sound of Pedgiree tbh, sounds like some sort of fancy cheese :D

**Review by daniaki — APPROVED** [review]
Can't see any obvious issues in `tests/`, and they're passing which is great 🎉 

Changes in the other files assigned to me also look good.

**Review by silkm — APPROVED** [review]
config files look all good. Readme has one use of sample in the example code [here](https://github.com/populationgenomics/production-pipelines/pull/366/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R681), does this need to be changed? Otherwise all good!

File: cpg_workflows/jobs/validation.py [hunk 2]
```diff
@@ -223,20 +215,15 @@ def parse_and_post_results(
...
-        analysis_model=AnalysisModel(
-            sample_ids=[sample.id],
-            type=AnalysisType('qc'),
-            status=AnalysisStatus('completed'),
-            output=str(happy_csv.parent),
-            meta=summary_data,
-            active=True,
-        ),
+    # NOTE: This change is untested.
+    get_metamist().create_analysis(
```
Comment (MattWellie) [inline cpg_workflows/jobs/validation.py:222]: @vivbak just checking, is this new method version already available?

**Review by MattWellie — APPROVED** [review]
Looks good! Heroic effort

File: cpg_workflows/jobs/validation.py (hunk 2 above)
Comment (vivbak) [reply cpg_workflows/jobs/validation.py:222]: Do you mean the updated create_analysis endpoint in metamist?

File: cpg_workflows/targets.py [hunk 3]
```diff
@@ -493,7 +513,7 @@ def __init__(
         sex: Sex | None = None,
         pedigree: Optional['PedigreeInfo'] = None,
         alignment_input_by_seq_type: dict[str, AlignmentInput] | None = None,
-        seq_by_type: dict[str, Sequence] | None = None,
+        assays: dict[str, Assay] | None = None,
```
Comment (illusional) [inline cpg_workflows/targets.py:516]: Is this keyed by external id, otherwise why a string?

File: cpg_workflows/targets.py [hunk 4]
```diff
@@ -512,7 +532,7 @@ def __init__(
         self.alignment_input_by_seq_type: dict[str, AlignmentInput] = (
             alignment_input_by_seq_type or dict()
         )
-        self.seq_by_type: dict[str, Sequence] = seq_by_type or dict()
+        self.assays: dict[str, Assay] = assays or dict()
```
Comment (illusional) [inline cpg_workflows/targets.py:535]: Not a big deal, but I thought pylint would complain about this
```suggestion
        self.assays: dict[str, Assay] = assays or {}
```

**Review by illusional — APPROVED** [review]
My section looks good! Few small comments, but overall amazing work 🎉

File: cpg_workflows/utils.py [hunk 5]
```diff
@@ -145,16 +145,16 @@ def slugify(line: str):
...
     file_names: list[str | ResourceFile],
 ) -> str:
     """
-    Helper function to add seds into a command that would extend samples IDs
+    Helper function to add seds into a command that would extend sequencing group IDs
     in each file in `file_names` with an external ID, only if external ID is
     different from the original.
 
-    @param rich_id_map: map used to replace samples, e.g. {'CPG1': 'CPG1|EXTID'}
+    @param rich_id_map: map used to replace sequencing groups, e.g. {'CPG1': 'CPG1|EXTID'}
```
Comment (jmarshall) [inline cpg_workflows/utils.py:157]: Loop variables below could be `sgid, rich_sgid` if anyone can be bothered.

File: cpg_workflows/workflow.py [hunk 6]
```diff
@@ -383,7 +391,7 @@ class Action(Enum):
 class Stage(Generic[TargetT], ABC):
     """
     Abstract class for a workflow stage. Parametrised by specific Target subclass,
-    i.e. SampleStage(Stage[Sample]) should only be able to work on Sample(Target).
+    i.e. SequencingGroup(Stage[SequencingGroup]) should only be able to work on SequencingGroup(Target).
```
Comment (jmarshall) [inline cpg_workflows/workflow.py:394]:
```suggestion
    i.e. SequencingGroupStage(Stage[SequencingGroup]) should only be able to work on SequencingGroup(Target).
```

File: cpg_workflows/inputs.py [hunk 7]
```diff
@@ -35,187 +35,119 @@ def create_cohort() -> Cohort:
...
 
 
 def _populate_analysis(cohort: Cohort) -> None:
     """
     Populate Analysis entries.
     """
     for dataset in cohort.get_datasets():
-        gvcf_by_sid = get_metamist().get_analyses_by_sid(
-            dataset.get_sample_ids(),
+        gvcf_by_sgid = get_metamist().get_analyses_by_sid(
```
Comment (jmarshall) [inline cpg_workflows/inputs.py:112]: Were these functions going to change to `get_…_by_sgid()`? (I guess the other two have evaporated, so only `get_analyses_by_s[g]id()` remains.)

File: cpg_workflows/jobs/genotype.py [hunk 8]
```diff
@@ -267,8 +267,8 @@ def merge_gvcfs_job(
     j.storage(f'{len(gvcf_groups) * 1.5 + 2}G')
     j.declare_resource_group(
         output_gvcf={
-            'g.vcf.gz': '{root}-' + sample_name + '.g.vcf.gz',
-            'g.vcf.gz.tbi': '{root}-' + sample_name + '.g.vcf.gz.tbi',
+            'g.vcf.gz': '{root}-' + sequencing_group_name + '.g.vcf.gz',
+            'g.vcf.gz.tbi': '{root}-' + sequencing_group_name + '.g.vcf.gz.tbi',
```
[the remark this answers is not in the document]
Comment (jmarshall) [reply cpg_workflows/jobs/genotype.py:271]: Suggestion should have been `'…': f'…'`, but 🤷 either way.

File: cpg_workflows/metamist.py [hunk 9]
```diff
@@ -32,6 +30,76 @@
     FastqPairs,
 )
 
+GET_SEQUENCNG_GROUPS_QUERY = gql(
```
[the remark this answers is not in the document]
Comment (vivbak) [reply cpg_workflows/metamist.py:33]: Thank you!

File: cpg_workflows/metamist.py [hunk 10]
```diff
@@ -153,111 +221,58 @@ class Metamist:
...
-            self.aapi.update_analysis_status(
+            self.aapi.update_analysis(
                 analysis.id,
                 models.AnalysisUpdateModel(status=models.AnalysisStatus(status.value)),
             )
         except ApiException:
             traceback.print_exc()
         analysis.status = status
 
+    # NOTE: This isn't used anywhere.
```
[the remark this answers is not in the document]
Comment (vivbak) [reply cpg_workflows/metamist.py:268]: I deemed this is an 'outside of scope' change, but left a note so it wasn't forgotten.

File: cpg_workflows/metamist.py [hunk 11]
```diff
@@ -274,54 +289,49 @@ def find_joint_calling_analysis(
             return None
         assert a.type == AnalysisType.JOINT_CALLING, data
         assert a.status == AnalysisStatus.COMPLETED, data
-        if a.sample_ids != set(sample_ids):
+        if a.sequencing_group_ids != set(sequencing_group_ids):
             return None
         return a
 
     def get_analyses_by_sid(
```
[the remark this answers is not in the document]
Comment (vivbak) [reply cpg_workflows/metamist.py:296]: Good catch

File: cpg_workflows/metamist.py [hunk 12]
```diff
@@ -473,38 +498,21 @@ def get_ped_entries(self, dataset: str | None = None) -> list[dict[str, str]]:
         if get_config()['workflow']['access_level'] == 'test':
             metamist_proj += '-test'
 
-        entries = self.fapi.get_families(metamist_proj)
-        family_ids = [entry['id'] for entry in entries]
+        family_ids = self.get_family_ids(dataset=dataset)
```
[the remark this answers is not in the document]
Comment (vivbak) [reply cpg_workflows/metamist.py:501]: oh gosh, so true!

[the remark this answers is not in the document]
Comment (vivbak) [reply cpg_workflows/metamist.py]: Scrapped a bunch of this

File: cpg_workflows/jobs/validation.py (hunk 2 above)
Comment (MattWellie) [reply cpg_workflows/jobs/validation.py:222]: Yup! AFAIK this is a renamed method in the `metamist` library, so this is currently available 👌

File: scripts/populate_metamist.py [hunk 13]
```diff
@@ -40,8 +40,8 @@ def main(command: str, config_paths: list[str]):
     status = MetamistStatusReporter()
 
     if command == 'analyses':
-        from sample_metadata.apis import AnalysisApi
-        from sample_metadata.models import AnalysisQueryModel, AnalysisStatus
+        from metamist.apis import AnalysisApi
+        from metamist.models import AnalysisQueryModel, AnalysisStatus
```
[the remark this answers is not in the document]
Comment (vivbak) [reply scripts/populate_metamist.py:44]: for sure, I still don't really know why we do this in some places and when it makes sense to do so.

File: cpg_workflows/large_cohort/combiner.py [hunk 14]
```diff
@@ -7,47 +7,45 @@
...
+            if get_config()['workflow'].get('skip_sgs_with_missing_input', False):
+                logging.warning(f'Skipping {sequencing_group} which is missing GVCF')
+                sequencing_group.active = False
                 continue
             else:
                 raise ValueError(
-                    f'Sample {sample} is missing GVCF. '
-                    f'Use workflow/skip_samples = [] or '
-                    f'workflow/skip_samples_with_missing_input '
+                    f'SequencingGroup {sequencing_group} is missing GVCF. '
```
Comment (jmarshall) [inline cpg_workflows/large_cohort/combiner.py:28]:
```suggestion
                    f'Sequencing group {sequencing_group} is missing GVCF. '
```

File: cpg_workflows/large_cohort/combiner.py [hunk 15]
```diff
@@ -7,47 +7,45 @@
...
+                        f'Skipping {sequencing_group} that is missing GVCF {sequencing_group.gvcf.path}'
                     )
-                    sample.active = False
+                    sequencing_group.active = False
                 else:
                     raise ValueError(
-                        f'Sample {sample} is missing GVCF. '
-                        f'Use workflow/skip_samples = [] or '
-                        f'workflow/skip_samples_with_missing_input '
+                        f'SequencingGroup {sequencing_group} is missing GVCF. '
```
Comment (jmarshall) [inline cpg_workflows/large_cohort/combiner.py:43]:
```suggestion
                        f'Sequencing group {sequencing_group} is missing GVCF. '
```

File: cpg_workflows/inputs.py [hunk 16]
```diff
@@ -225,29 +157,25 @@ def _populate_pedigree(cohort: Cohort) -> None:
...
-            maternal_sample = sample_by_participant_id.get(
-                str(ped_entry['maternal_id'])
-            )
-            paternal_sample = sample_by_participant_id.get(
-                str(ped_entry['paternal_id'])
-            )
-            sample.pedigree = PedigreeInfo(
-                sample=sample,
+            ped_entry = ped_entry_by_participant_id[sequencing_group.participant_id]
+            maternal_sample = sg_by_participant_id.get(str(ped_entry['maternal_id']))
```
Comment (jmarshall) [inline cpg_workflows/inputs.py:167]: `maternal_sg` maybe 🤷

File: cpg_workflows/stages/gvcf_qc.py [hunk 17]
```diff
@@ -26,82 +26,90 @@
...
         """
         Parsed by MultiQC: '*.summary.csv'
         https://multiqc.info/docs/#hap.py
         """
-        if sample.participant_id not in get_config().get('validation', {}).get(
-            'sample_map', {}
-        ):
+        if sequencing_group.participant_id not in get_config().get(
+            'validation', {}
+        ).get('sample_map', {}):
```
Comment (jmarshall) [inline cpg_workflows/stages/gvcf_qc.py:82]: Are we leaving this configuration key as `sample_map`?

File: cpg_workflows/stages/joint_genotyping_qc.py [hunk 18]
```diff
@@ -69,53 +69,57 @@ def queue_jobs(self, cohort: Cohort, inputs: StageInput) -> StageOutput | None:
...
         """
         Parsed by MultiQC: '*.summary.csv'
         https://multiqc.info/docs/#hap.py
         """
-        if sample.participant_id not in get_config().get('validation', {}).get(
-            'sample_map', {}
-        ):
+        if sequencing_group.participant_id not in get_config().get(
+            'validation', {}
+        ).get('sample_map', {}):
```
Comment (jmarshall) [inline cpg_workflows/stages/joint_genotyping_qc.py:84]: As per _gvcf_qc.py_: are we leaving this configuration key as `sample_map`?

**Review by jmarshall — APPROVED** [review]
I've reviewed all the files assigned to me, and that's all the comments I've got. I'm not really qualified to review the metamist-change-related parts of _inputs.py_ (in `create_cohort` and `_populate_alignment_inputs`) but have looked at the rest of that file.

There's a certain amount of `/sample/` remaining in _large_cohort/sample_qc.py_ but that is probably fine.

I think in due course and for new code we will want to shorten local variables as `seqgroup` or (when very local) `sg` or something rather than the rather unwieldy `sequence_group`, but let's not delay by thinking about that now!

Approved, with several comments that may or may not be worth addressing.

File: cpg_workflows/targets.py (hunk 3 above)
Comment (vivbak) [reply cpg_workflows/targets.py:516]: It's keyed by assay-type, which AFAIK is a string, I can define an enum for it though if that's better? It's meant to be the direct replacement of seq_by_type, so I can alternatively call it assays_by_type

File: cpg_workflows/stages/gvcf_qc.py (hunk 17 above)
Comment (vivbak) [reply cpg_workflows/stages/gvcf_qc.py:82]: I've left this for now, as it appears to be mapping to 'external sample ids'.

File: cpg_workflows/stages/joint_genotyping_qc.py (hunk 18 above)
Comment (vivbak) [reply cpg_workflows/stages/joint_genotyping_qc.py:84]: As above

File: cpg_workflows/stages/gvcf_qc.py (hunk 17 above)
Comment (jmarshall) [reply cpg_workflows/stages/gvcf_qc.py:82]: Fair enough, makes sense.

**Pull request merged**